node/src/daser.rs

```diff
@@ -0,0 +1,309 @@
+
+        spawn(async move {
+            if let Err(e) = worker.run().await {
+                error!("Fatal DASer error: {e}");
+            }
+        });
+
+        Ok(Daser { cancellation_token })
+    }
+
+    /// Stop the [`Daser`].
+    pub fn stop(&self) {
+        // Singal the Worker to stop.
+        // TODO: Should we wait for the Worker to stop?
+        self.cancellation_token.cancel();
+    }
+}
+
+impl Drop for Daser {
+    fn drop(&mut self) {
+        self.cancellation_token.cancel();
+    }
+}
+
+struct Worker<S>
+where
+    S: Store + 'static,
+{
+    cancellation_token: CancellationToken,
+    p2p: Arc<P2p>,
+    store: Arc<S>,
+    max_samples_needed: usize,
+}
+
+impl<S> Worker<S>
+where
+    S: Store,
+{
+    fn new(args: DaserArgs<S>, cancellation_token: CancellationToken) -> Result<Worker<S>> {
+        Ok(Worker {
+            cancellation_token,
+            p2p: args.p2p,
+            store: args.store,
+            max_samples_needed: MAX_SAMPLES_NEEDED,
+        })
+    }
+
+    async fn run(&mut self) -> Result<()> {
+        let cancellation_token = self.cancellation_token.clone();
+
+        loop {
+            select! {
+                _ = cancellation_token.cancelled() => break,
+                res = self.sample_next_block() => res?,
+            }
+        }
+
+        Ok(())
+    }
+
+    async fn sample_next_block(&mut self) -> Result<()> {
+        let height = self.store.next_unsampled_height().await?;
+        self.store.wait_height(height).await?;
+
+        let header = self.store.get_by_height(height).await?;
+        let (sampled_cids, accepted) = self.sample_block(&header).await?;
+
+        self.store
+            .update_sampling_metadata(height, accepted, sampled_cids)
+            .await?;
+
+        Ok(())
+    }
+
+    async fn sample_block(&mut self, header: &ExtendedHeader) -> Result<(Vec<Cid>, bool)> {
+        let now = Instant::now();
+        let block_len = header.dah.square_len() * header.dah.square_len();
+        let indexes = random_indexes(block_len, self.max_samples_needed);
+        let mut futs = FuturesUnordered::new();
+
+        for index in indexes {
+            let fut = self
+                .p2p
+                .get_sample(index, header.dah.square_len(), header.height().value());
+            futs.push(fut);
+        }
+
+        let mut cids: Vec<Cid> = Vec::new();
+        let mut accepted = true;
+
+        while let Some(res) = futs.next().await {
+            match res {
+                Ok(sample) => cids.push(convert_cid(&sample.sample_id.into())?),
+                // Validation is done in Bitswap level thought `ShwapMultihasher`.
+                // When a sample is not valid, then is will never be delivered
+                // to us as the data of the CID. Because of that, the only way
+                // to know that data sampling verification failed is when a query
+                // times out.
```

Comment:
```suggestion
                // Validation is done at Bitswap level, though `ShwapMultihasher`.
                // If the sample is not valid, it will never be delivered to us
                // as the data of the CID. Because of that, the only signal 
                // that data sampling verification failed is query timing out.
```

Reply:
typo in suggestion, 'through'